posts/2023-09-21-closer-look-private-members.md

```diff
@@ -0,0 +1,89 @@
+
+While YAML is very flexible and the serializer currently uses reflection to set the properties, the runtime serialization is done using a custom binary serializer system. When you compile your game project, after C# compiler has done its job, the build system runs the AssemblyProcessor over your code to generate efficient serializer classes that do not use reflection to access members on the objects they deserialize.
+
+Access modifiers are enforced by the CLR, which means it's not possible to generate code that accesses private members of another class directly.
+
+### What about constructors?
+
+If we have a private field that we want to set once, why don't we use a constructor to accept a value for it? It's certainly an option and there are serialization systems which look at constructors to provide deserialization in this way. They may be a bit more complicated but it's possible to implement them.
+
+However, in a generic environment that Stride provides it's a bit difficult. How would the editor create a new instance of your class if it requires data to be passed in the constructor? Will it provide `default` values, will it look for a private parameterless constructor?
+
+### What about init properties?
+
+C# 9 has introduced [`init`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/init) properties which are only allowed to be set at object construction. Reflection is able to bypass that of course, while at the CLR level a special marker `modreq` is placed on the property which asks the compiler to provide `modreq` marker on the callsite as well (see [thread](https://stackoverflow.com/a/64749884)). Compilers which don't know about init properties would still prevent you from assigning them. And the CLR is not actually enforcing anything here - only compiler - thus the generated serializers are still able to use those properties.
+
+But in your code the constraint will be enforced and no runtime modification will occur after the object is initialized.
+
+```csharp
+public string MyProperty { get; init; }
+```
+
+### `UnsafeAccessor` in .NET 8
+
+As we're trying to squeeze out more performance and improve the Stride codebase we're looking at rewriting the AssemblyProcessor from Mono.Cecil and manually emitting IL into a Roslyn C# source generator. This will also mean being restricted to whatever the compiler enforces, such as not setting init properties outside of object construction.
+
+Luckily in .NET 8 there's a new performance hack which allows you to access private fields on other classes. You can read about it more in the [proposal](https://github.com/dotnet/runtime/issues/90081) and on the [performance blog](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/#networking-primitives).
```

Comment:
@VaclavElias I just realized I linked the wrong proposal link. Should be this one https://github.com/dotnet/runtime/issues/81741

Reply:
That is ok. I can update it as I am still testing from staging :)